extensions/amp-live-list/0.1/poller.js

```diff
@@ -0,0 +1,201 @@
+ */
+export class Poller {
+
+  constructor(win, wait, work) {
+    /** @private @const {!Window} */
+    this.win = win;
+
+    /** @private {number} */
+    this.wait_ = wait;
+
+    /** @private {function(): !Promise} */
+    this.work_ = work;
+
+    /** @private {?number} */
+    this.lastTimeoutId_ = null;
+
+    /** @private {?number} */
+    this.lastBackoffTimeoutId_ = null;
+
+    /** @private {?function(): number} */
+    this.backoffPoller_ = null;
+
+    /** @private {!PollerState} */
+    this.state_ = PollerState.STOPPED;
+
+    /**
+     * Mostly for testing purposes.
+     * @private {!Promise}
+     */
+    this.lastWorkPromise_ = null;
+  }
+
+  /**
+   * @return {function(): number}
+   * @private
+   */
+  getBackoffPoller_() {
+    return exponentialBackoff();
+  }
+
+  /**
+   * @return {!PollerState}
+   */
+  getState() {
+    return this.state_;
+  }
+
+  /**
+   * Initalize any work needed to start polling.
+   */
+  start() {
+    if (this.lastTimeoutId_) {
+      return;
+    }
+
+    if (this.lastBackoffTimeoutId_) {
+      this.win.clearTimeout(this.lastBackoffTimeoutId_);
+      this.lastBackoffTimeoutId_ = null;
+    }
+
+    // We reset the backoff poller
+    this.backoffPoller_ = null;
+    this.state_ = PollerState.RUNNING;
+    this.poll_();
+  }
+
+  /**
+   * Clear timers and set correct stop state.
+   */
+  stop() {
+    if (this.state_ == PollerState.STOPPED) {
+      return;
+    }
+
+    this.state_ = PollerState.STOPPED;
+    if (this.lastTimeoutId_) {
+      this.win.clearTimeout(this.lastTimeoutId_);
+      this.lastTimeoutId_ = null;
+    }
+    if (this.lastBackoffTimeoutId_) {
+      this.win.clearTimeout(this.lastBackoffTimeoutId_);
+      this.lastBackoffTimeoutId_ = null;
+      this.backoffPoller_ = null;
+    }
+  }
+
+  /**
+   * Assign correct state and call internal recursive backoffPoll_.
+   * @private
+   */
+  backoff_() {
+    if (this.lastBackoffTimeoutId_) {
+      return;
+    }
+    if (!this.backoffPoller_) {
+      this.backoffPoller_ = this.getBackoffPoller_();
+    }
+    this.stop();
+
+    this.state_ = PollerState.BACKEDOFF;
+    this.backoffPoll_();
+  }
+
+  /**
+   * @private
+   */
+  backoffPoll_() {
+    if (this.state_ != PollerState.BACKEDOFF) {
+      return;
+    }
+    this.lastBackoffTimeoutId_ = this.backoffPoller_(() => {
+      this.lastWorkPromise_ = this.work_().then(this.start.bind(this))
+          .catch((err) => {
+            if (err.retriable) {
+              this.backoffPoll_();
+            } else {
+              throw err;
+            }
+          });
+    });
+  }
+
+  /**
+   * @private
+   */
+  poll_() {
+    if (this.state_ != PollerState.RUNNING) {
+      return;
+    }
+
+    const wait = Poller.getWaitTime_(this.wait_);
+    this.lastTimeoutId_ = this.win.setTimeout(() => {
+      this.lastWorkPromise_ = this.work_().then(this.poll_.bind(this))
+          .catch((err) => {
+            if (err.retriable) {
```

Comment:
I think 304 will actually yield an empty response as far as XHR concerned. We discussed with @erwinmombay and will definitely test this case. Most likely we will change `xhr.js` API to return `null` for document in case of 304.